logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java

```diff
@@ -139,88 +231,54 @@ public static Dataset terminalDataset(final Collection<Dataset> parents) {
-                syntax.append(clear(i));
-            }
+            clearSyntax = "";
+            inlineClear = clearSyntax(offset);
         } else {
-            for (int i = 0; i < cnt; ++i) {
-                clearSyntax.append(clear(i));
-            }
+            inlineClear = "";
+            clearSyntax = clearSyntax(offset);
         }
-        syntax.append(RETURN_NULL);
-        return compile(syntax.toString(), clearSyntax.toString(), allArgs);
-    }
-
-    /**
-     * Special case optimization for when the output plugin is directly connected to the Queue
-     * without any filters or conditionals in between. This special case does not arise naturally
-     * from {@link DatasetCompiler#outputDataset(Collection, IRubyObject, boolean)} since it saves
-     * the internal buffering of events and instead forwards events directly from the batch to the
-     * Output plugin.
-     * @param output Output Plugin
-     * @return Dataset representing the Output
-     */
-    private static Dataset outputDatasetFromRoot(final IRubyObject output,
-        final DynamicMethod method) {
-        final int argArrayIndex = 1;
-        final StringBuilder syntax = new StringBuilder();
-        syntax.append(field(argArrayIndex)).append("[0] = batch;");
-        final int callsiteIndex = 0;
-        final int pluginIndex = 2;
-        syntax.append(callOutput(callsiteIndex, argArrayIndex, pluginIndex));
-        final Object[] allArgs = new Object[3];
-        allArgs[callsiteIndex] = method;
-        allArgs[argArrayIndex] = new IRubyObject[1];
-        allArgs[pluginIndex] = output;
-        syntax.append(RETURN_NULL);
-        return compile(syntax.toString(), "", allArgs);
+        return compileOutput(
+            join(
+                join(
+                    bufferForOutput(offset), callOutput(offset), clear(offset + INPUT_BUFFER_OFFSET)
+                ), inlineClear
+            ), clearSyntax, allArgs
+        );
     }
 
-    /**
-     * Generates the code for invoking the Output plugin's `multi_receive` method.
-     * @param callsiteIndex Field index of the `multi_receive` call site
-     * @param argArrayIndex Field index of the invocation argument array
-     * @param pluginIndex Field index of the Output plugin's Ruby object
-     * @return Java Code String
-     */
-    private static String callOutput(final int callsiteIndex, final int argArrayIndex,
-        final int pluginIndex) {
-        return new StringBuilder().append(field(callsiteIndex)).append(
-            ".call(RubyUtil.RUBY.getCurrentContext(), ").append(field(pluginIndex))
-            .append(", RubyUtil.LOGSTASH_MODULE, \"multi_receive\", ")
-            .append(field(argArrayIndex)).append(", Block.NULL_BLOCK);").toString();
+    private static String callFilterFlush(final int offset, final boolean shutdownOnly) {
+        final String condition;
+        final String flushArgs;
+        if (shutdownOnly) {
+            condition = "flush && shutdown";
```

Comment:
nit: use the flush constant above, and create one for shutdown.

Reply:
@jakelandis hmm I thought the same :D, but:

* this `flush` is a method parameter, the constant was using in place of the method name when invoking a Ruby call-site. I think reusing it here would make things more confusing since this spot (at least syntax-wise) has nothing to do with other uses of it.

That said ... you're right, this shouldn't have some magic `String` duplication to it. How about renaming the `flush` param to `doFlush` throughout the codebase (would just be in `Dataset` code)?